larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -532,206 +533,115 @@ void StitchingCosmicRayMergingTool::StitchPfos(const MasterAlgorithm *const pAlg
+                if (!LArStitchingHelper::CanTPCsBeStitched(*pLArTPCI, *pLArTPCJ))
+                    continue;
 
-            if ((iter1 != pfoToLArTPCMap.end()) && (iter2 != pfoToLArTPCMap.end()))
-            {
-                const LArTPC *const pLArTPC1(iter1->second);
-                const LArTPC *const pLArTPC2(iter2->second);
+                PfoList pfosToShift({pPfoI, pPfoJ});
 
-                if (((twoDHitsPfo1 + twoDHitsPfo2) > totalHitsStitched) && LArStitchingHelper::CanTPCsBeStitched(*pLArTPC1, *pLArTPC2))
+                for (const ParticleFlowObject *const pPfoToShift : pfosToShift)
                 {
-                    totalHitsStitched = twoDHitsPfo1 + twoDHitsPfo2;
-                    reducedPfoVector.clear();
-                    reducedPfoVector.push_back(pPfoToShift1);
-                    reducedPfoVector.push_back(pPfoToShift2);
+                    // if pfo has already been shifted
+                    if (std::find(shiftedPfos.begin(), shiftedPfos.end(), pPfoToShift) != shiftedPfos.end())
+                        continue;
+
+                    if (!m_useXcoordinate || m_alwaysApplyT0Calculation)
+                    {
+                        // get pfo stitching vertex
+                        const PfoToPointingVertexMatrix::iterator pfoToPointingVertexMatrixIter(pfoToPointingVertexMatrix.find(pPfoToShift));
+                        LArPointingCluster::Vertex stitchingVertex;
+                        
+                        for (const ParticleFlowObject *const pPfo : pfosToShift)
+                        {
+                            if (pPfo == pPfoToShift)
+                                continue;
+
+                            stitchingVertex = pfoToPointingVertexMatrixIter->second.at(pPfo);
+                        }
+
+                        // Determine shift sign from the relative position of stitching vertex and closest TPC boundary position
+                        const LArTPC *const pShiftLArTPC(pfoToLArTPCMap.at(pPfoToShift));
+                        const float tpcBoundaryCenterX(LArStitchingHelper::GetTPCBoundaryCenterX(*pLArTPCI, *pLArTPCJ));
+                        float tpcBoundaryX(0.f);
+
+                        if (pShiftLArTPC->GetCenterX() < tpcBoundaryCenterX)
+                        {
+                            tpcBoundaryX = pShiftLArTPC->GetCenterX() + (pShiftLArTPC->GetWidthX()/2);
+                        }
+                        else
+                        {
+                            tpcBoundaryX = pShiftLArTPC->GetCenterX() - (pShiftLArTPC->GetWidthX()/2);
+                        }
+
+                        const float positionShiftSign = stitchingVertex.GetPosition().GetX() < tpcBoundaryX ? 1.f : -1.f;
+
+                        // ATTN: No CPA/APA sign needed since x0 calculation is corresponds to an APA
+                        object_creation::ParticleFlowObject::Metadata metadata;
+                        metadata.m_propertiesToAdd["X0"] = x0;
```

Comment:
After lots of head scratching, yes. The CalculateX0 function now calculates an average for each TPC boundary. If it is an APA boundary, the contribution to the average is just the X0 calculation and if it is a CPA it is multiplied by -1. For an APA when muons arrive before the beam they have a negative X0 and if they arrive after they have a positive X0 (therefore opposite for the CPA). So in terms of time, if they arrive before the beam they have a negative time and if they arrive after they have a positive sign. The sign of the time is maintained when we have a CPA boundary because of this sign multiplication in the CalculateX0 function. When I made these plots I filled a tree with the exact variables at the same position in the code where the map was filled, I also checked these. However, I should have accessed it through the metadata, although I couldn't work out how to do this at the time..

Reply:
This sounds quite clear and convincing to me (especially given the complexity of the problem!). It would be nice to add some (maybe more terse) description along these lines in the function doxygen brief. Thanks!